src/net_processing.cpp

```diff
@@ -1264,6 +1264,9 @@ bool static ProcessHeadersMessage(CNode *pfrom, CConnman *connman, const std::ve
             if (nDoS > 0) {
                 LOCK(cs_main);
                 Misbehaving(pfrom->GetId(), nDoS);
+            } else if (!pfrom->fInbound && !pfrom->m_manual_connection && punish_invalid) {
```

Comment:
Carrying forward my cranky-old-man-rant from: https://github.com/bitcoin/bitcoin/pull/11446#discussion_r144654182

This is crazy-confusing to read. If the header was bad enough to earn them some ban points, put the points on the score-board and carry on. But if it wasn't bad enough for points, kick them immediately.

If we're punishing for invalid, I think the disconnect needs to always happen ```if (state.IsInvalid)```. Using points as a proxy for some certain invalid condition(s) is really non-obvious.

Reply:
I agree with @theuni. Think this should be `if ...` not `else if ...` so that we always disconnect on invalid blocks.

Reply:
Obviously this is very, very much a stop-gap for 0.15.0.2. Next up is refactoring the CValidationState logic to have a concept of *what* went wrong (bad-by-soft-fork, bad-by-prev-block, bad-by-....) and then just move all of the DoS logic into net_processing. I believe a switch to if instead of else if may break the "prev block not found" handling, so I'm happy to commit to pushing for a DoS overhaul by 0.16 in the coming months.

Reply:
After further analysis, I think the logic I've presented here is undesirable, because there is at least one situation where we should not disconnect an outbound peer for an invalid block header, which is when the block timestamp is too far in the future.  This is a problem whether or not we break this `else if` out into it's own `if` clause.

I'm going to add a commit that reworks this to explicitly look for the case where the first invalid block header processed is one for which we have the entry in our mapBlockIndex, and only disconnect in that case.  I think the review burden will be somewhat (much?) lower for such an approach, and it still accomplishes the overall goal of disconnecting peers on (some, but not all) known-invalid headers chains.